### ISSUE TITLE
TestEngine with ID 'junit-vintage' failed to discover tests. NoClassDefFoundError: org/bukkit/scoreboard/Criteria

### ISSUE THREAD

**Issue by escape9179** [MockBukkit/MockBukkit#623]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Are you using the latest version of MockBukkit?

- [X] I am using the latest version of MockBukkit.

### Minecraft Version

1.19

### Describe the bug

When attempting to run all tests in a directory I get this error:
```
Internal Error occurred.
org.junit.platform.commons.JUnitException: TestEngine with ID 'junit-vintage' failed to discover tests
	at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discoverEngineRoot(EngineDiscoveryOrchestrator.java:160)
	at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discoverSafely(EngineDiscoveryOrchestrator.java:132)
	at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discover(EngineDiscoveryOrchestrator.java:107)
	at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discover(EngineDiscoveryOrchestrator.java:78)
	at org.junit.platform.launcher.core.DefaultLauncher.discover(DefaultLauncher.java:110)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:86)
	at org.junit.platform.launcher.core.DefaultLauncherSession$DelegatingLauncher.execute(DefaultLauncherSession.java:86)
	at org.junit.platform.launcher.core.SessionPerRequestLauncher.execute(SessionPerRequestLauncher.java:53)
	at com.intellij.junit5.JUnit5IdeaTestRunner.startRunnerWithArgs(JUnit5IdeaTestRunner.java:57)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater$1.execute(IdeaTestRunner.java:38)
	at com.intellij.rt.execution.junit.TestsRepeater.repeat(TestsRepeater.java:11)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:35)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:235)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:54)
Caused by: java.lang.NoClassDefFoundError: org/bukkit/scoreboard/Criteria
	at java.base/java.lang.Class.getDeclaredMethods0(Native Method)
	at java.base/java.lang.Class.privateGetDeclaredMethods(Class.java:3434)
	at java.base/java.lang.Class.privateGetPublicMethods(Class.java:3459)
	at java.base/java.lang.Class.privateGetPublicMethods(Class.java:3465)
	at java.base/java.lang.Class.getMethods(Class.java:2051)
	at org.junit.platform.commons.util.ReflectionUtils.getDefaultMethods(ReflectionUtils.java:1520)
	at org.junit.platform.commons.util.ReflectionUtils.getDeclaredMethods(ReflectionUtils.java:1493)
	at org.junit.platform.commons.util.ReflectionUtils.findAllMethodsInHierarchy(ReflectionUtils.java:1435)
	at org.junit.platform.commons.util.ReflectionUtils.findMethods(ReflectionUtils.java:1419)
	at org.junit.platform.commons.util.ReflectionUtils.findMethods(ReflectionUtils.java:1405)
	at org.junit.vintage.engine.descriptor.TestSourceProvider.lambda$findMethod$1(TestSourceProvider.java:75)
	at java.base/java.util.HashMap.computeIfAbsent(HashMap.java:1220)
	at java.base/java.util.Collections$SynchronizedMap.computeIfAbsent(Collections.java:2760)
	at org.junit.vintage.engine.descriptor.TestSourceProvider.findMethod(TestSourceProvider.java:75)
	at org.junit.vintage.engine.descriptor.TestSourceProvider.computeTestSource(TestSourceProvider.java:56)
	at java.base/java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1708)
	at org.junit.vintage.engine.descriptor.TestSourceProvider.findTestSource(TestSourceProvider.java:47)
	at org.junit.vintage.engine.discovery.RunnerTestDescriptorPostProcessor.addChildrenRecursively(RunnerTestDescriptorPostProcessor.java:62)
	at org.junit.vintage.engine.discovery.RunnerTestDescriptorPostProcessor.applyFiltersAndCreateDescendants(RunnerTestDescriptorPostProcessor.java:41)
	at org.junit.vintage.engine.discovery.VintageDiscoverer.discover(VintageDiscoverer.java:46)
	at org.junit.vintage.engine.VintageTestEngine.discover(VintageTestEngine.java:64)
	at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discoverEngineRoot(EngineDiscoveryOrchestrator.java:152)
	... 13 more
Caused by: java.lang.ClassNotFoundException: org.bukkit.scoreboard.Criteria
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:521)
	... 35 more
```

Everything works normally when instead of clicking on a test directory and selecting an action to run all tests in that directory, I click on a specific test class and click on an action to run the tests in that class. I only get this error when I click on a directory that contains more than one test class and right-click the directory to perform an action that runs all tests within that directory.

**I don't get this error when using MockBukkit-v1.19 version 2.29.0. Everything works well. I haven't tested any versions between 2.29.0 and 2.117.1 to narrow down the possible version that's causing this issue.**

**Edit: I managed to narrow down the suspect. Version 2.111.7 is the cause of the error. Everything before 2.111.7 works with no error.**

**IDE:**
IntelliJ 2022.1 (IU-222.3739.54)

**Dependencies:**
<img width="564" alt="image" src="https://user-images.githubusercontent.com/9893025/188548166-e38358b7-6f54-40a8-8978-07d3cf7d6bd6.png">

**Directory Structure:**
<img width="207" alt="image" src="https://user-images.githubusercontent.com/9893025/188548297-9afb6be9-c785-4321-b269-c74a31606c1a.png">

**What's being clicked:**
<img width="410" alt="image" src="https://user-images.githubusercontent.com/9893025/188548366-b2cc0eb6-a0e0-4cf8-97da-ceffc2cdd4ae.png">

### Reproducible Test

Right-cilcking on a test class doesn't seem to produce any errors. Only when clicking on a test directory and clicking on an action to run all tests within that directory.

### Anything else?

As mentioned above, when performing the same actions that produce this bug while using MockBukkitv1.19 version 2.29.0, everything works fine. I get the error when using v2.117.1

**Comment by Insprill** [comment]
Interesting bug. Can you try to narrow down the version where this error appeared? You can do a sort of "binary search" by halving the version number until you get the issue to speed up the process.

**Comment by escape9179** [comment]
> Interesting bug. Can you try to narrow down the version where this error appeared? You can do a sort of "binary search" by halving the version number until you get the issue to speed up the process.

Narrowed it down to 2.111.7.

**Comment by Insprill** [comment]
Interesting. Some new scoreboard-related methods were stubbed with #606 which is that version. Have you tried invalidating caches? It's most likely IntelliJ not updating something.

**Comment by escape9179** [comment]
I've tried invalidating caches and restarting, but it didn't solve the issue. Maybe I'll try writing these tests in Java to see it has anything to do with Kotlin, or MockK.

**Comment by Insprill** [comment]
Can you try using [JUnit 5](https://junit.org/junit5/docs/current/user-guide/) with the [jupiter](https://mvnrepository.com/artifact/org.junit.jupiter/junit-jupiter-engine) engine to see if that fixes it? You can also try it on a fresh project to see if it is some sort of caching issue. Considering it only breaks when running from a directory it's much more likely this is an IntelliJ/JUnit issue than a MockBukkit issue.

**Comment by escape9179** [comment]
Okay. I found the cause of the error. My project has a dependency on the Paper API v1.19.1, and a dependency on MockBukkit-v1.19 v2.117.2. After trying JUnit 5 instead of 4, with no success, I looked up other possible dependencies that JUnit 5 might require and stumbled upon its page of dependency metadata and noticed that the vintage engine was already packaged with JUnit because it was a dependency of JUnit. **I wasn't getting the error because of JUnit or any of its dependencies**. After finding the dependency metadata for JUnit and finding that the vintage engine dependency from JUnit was omitted (because it was already in my projects POM), I decided to search through other metadata and found out that the version of MockBukkit I was using had a dependency on the Paper API as well, but because I already had the Paper API as a dependency in my POM, the Paper API from MockBukkit was omitted. MockBukkit's version of the dependency was 1.19.2 whereas mine had a version of 1.19.1. As soon as I switched to Paper API v1.19.2 the problem disappeared. Apparently `org.bukkit.scoreboard.Criteria` was added in 1.19.2, which is the class I was getting `NoClassDefFoundError` for because I was using Paper 1.19.1.
Thanks for your help.
